readme
编译:
一、
  make #直接make默认是arm-none-linux-gnueabi 交叉编译工具

二、
  make CROSS_PREFIX=arm-linux  #指定交叉编译前缀
注意：
  arm-none-linux-gnueabi-gcc, 前缀为CROSS_PREFIX=arm-none-linux-gnueabi
  不要写为CROSS_PREFIX=arm-none-linux-gnueabi-, 会导致fdk-aac configure报错


rtmpdump/ChangeLog
RTMPDump
Copyright 2008-2009 Andrej Stepanchuk; Distributed under the GPL v2
Copyright 2009-2011 Howard Chu
Copyright 2009 The Flvstreamer Team
http://rtmpdump.mplayerhq.hu/

11 July 2011, v2.4
- add RTMPE type 9 handshake support

30 June 2010, v2.3
- fix RC4 cleanup for GnuTLS/gcrypt
- declare RTMP_Write buf as const
- cleanup Makefile
- replace all use of bool with int
- add RTMP_Socket() and RTMP_Pause() APIs
- add ping/pong message handling
- add basic shared library support
- fix RTMP_ParseURL extension skipping
- fix bad switch() from -r477 in RTMP_Read
- fix rtmpsrv to always use .flv extension on output
- fix crash on socket failure while RTMP_Read() is reading header
- fix RTMP_ReadPacket signed/unsigned chars for m_nChannel
- license cleanup, use current FSF address
- fix RTMP_SetupURL tcUrl generation
- ignore multiple spaces between URL options
- only send CheckBW request once per session

29 May 2010, v2.2e
- port to Xbox
- add explicit URL scheme for rtmpts
- fix rtmpt clientID NUL-termination
- use BufferEmpty trick in rtmpdump but not by default in librtmp
- add librtmp manpage
- fix RTMP_Read, return 0 on EOF, not -1
- change RTMP_Read to return 1 packet at a time, fix buffer mgmt
- fix request/result queue, fix server compatibility for Publishing
- add RTMP_EnableWrite to enable Publishing

29 April 2010, v2.2d
- add RTMP_Alloc, RTMP_Free APIs
- add optional support for polarssl instead of OpenSSL
- add option to build crypto support without SSL/TLS
- tweak handshake offset checking
- add RTMP set_playlist command
- check for (and fix) broken timestamps in FLV packets
- fix tcUrl and playpath parsing in rtmpsrv and rtmpsuck
- change internal boolean flags to bitmasks

14 April 2010, v2.2c
- internal restructuring, fix #undef CRYPTO builds
- add RTMP_SetupURL, RTMP_SetOpt APIs
- add logging callback

22 March 2010, v2.2b
- fix v2.2a crashes in rtmpsrv/rtmpsuck
- fix v2.2a .swfinfo location on Windows
- fix typo for --auth parameter in manpages
- add FP10 handshake support for rtmpsrv/rtmpsuck
- avoid GNUMake vs BSDMake incompatibilities
- add pkgconfig file for librtmp
- more library cleanup

20 March 2010, v2.2a
- fix C++ compatibility for librtmp
- misc library restructuring
- add client support for tunneling: rtmpt, rtmpte, rtmps
- fix rtmpdump/rtmpgw FLV header dataType
- implement RTMP_Read() and RTMP_Write() to simplify library use
- fix SendPacket timestamps
- add optional support for GnuTLS/Gcrypt instead of OpenSSL
- use $HOMEPATH on Windows instead of $HOME for .swfinfo

4 March 2010, v2.2
- move RTMP code into library librtmp
- relicense RTMP code under LGPL v2.1
- add rtmpdump manpage
- fix AMF_LONG_STRING handling
- more FlashPlayer 10 handshake support
- in rtmpsrv fix Play Start/Stop messages
- rename "streams" program to "rtmpgw"

20 February 2010, v2.1d
- extend .swfinfo file format, add --swfAge rtmpdump parameter
  old file should be replaced or manually updated: copy the
  "date:" line and rename it to "ctim:"

- fix MacOSX builds - just use "make posix" now for all Unix-derived systems
- more explicit error checks in HTTP_get()
- in rtmpsrv spawn rtmpdump automatically
- fix bug in retry/resume of audio-only streams
- other minor misc. fixes

9 January 2010, v2.1c
- cleanup rtmpsrv output
- fix crash in 2.1b hashswf
- fix parseurl to url-decode PlayPath
- fix parseurl to recognize extensions followed by URL params
- fix Makefile, inadvertently dropped 'v' from version string
- in rtmpdump try Reconnect if ToggleStream doesn't work on timeouts
- in rtmpsuck use chunk-based I/O for better latency
- in rtmpsuck support lists of streams
- in rtmpsuck use raw client connect packet to workaround unsupported features
- support arbitrary AMF data appended to connect requests

4 January 2010, v2.1b
- fix url matching in .swfinfo lookup
- fix resume parsing in rtmpdump
- minor code cleanup (CRYPTO dependencies, logging)
- add getStreamLength recognition to rtmpsrv
- add close processing in rtmpsuck

1 January 2010, v2.1a
- fix socket receive timeouts for WIN32
- add streams description to README

29 December 2009, v2.1
- AMF cleanup: bounds checking for all encoders, moved AMF_EncodeNamed* from rtmp.c
- added SecureToken support
- added automatic SWF hash calculation
- added server-side handshake processing
- added rtmpsrv stub server example
- added rtmpsuck proxy server
- tweaks for logging
- renamed more functions to cleanup namespace for library use
- tweaks for server operation: objectEncoding, chunksize changes

16 December 2009, v2.0
- rewrote everything else in C, reorganized to make it usable again as a library
- fixed more portability bugs
- plugged memory leaks

2 December 2009, v1.9a
- fix auth string typo
- handle FCUnsubscribe message
- don't try retry on live streams
- SIGPIPE portability fix
- remove "not supported" comment for RTMPE

13 November 2009, v1.9
- Handle more signals to reduce risk of unresumable/corrupted partially streamed files
- Fixed >2GB file handling
- Added --hashes option for a hash progress bar instead of byte counter
- Fix to allow win32 to use binary mode on stdout.
- Added auto-unpause for buffer-limited streams

1 November 2009, v1.7

- added --subscribe option for subscribing to a stream
- added --start / --stop options for specifying endpoints of a stream
- added --debug / --quiet / --verbose options for controlling output

- added SOCKS4 support (by Monsieur Video)

- restructured to support auto-restart of timed-out streams

- rewritten byteswapping, works on all platforms

- fixed errors in command / result parsing

- support functions rewritten in C to avoid g++ compiler bugs on ARM

- support for 65600 channels instead of just 64

- fixed signature buffer overruns

17 May 2009, v1.6

- big endian alignment fix, should fix sparc64 and others

- moved timestamp handling into RTMP protocol innings, all packets have
absolute timestamps now, when seeking the stream will start with timestamp 0

rtmpsdk/g711.c
#include <stdlib.h>
#include <stdint.h>
#include "g711.h"

#define SIGN_BIT   (0x80)   /* Sign bit for a A-law byte. */
#define QUANT_MASK (0xf)    /* Quantization field mask. */
#define NSEGS      (8)      /* Number of A-law segments. */
#define SEG_SHIFT  (4)      /* Left shift for segment number. */
#define SEG_MASK   (0x70)   /* Segment field mask. */
#define BIAS       (0x84)   /* Bias for linear code. */


int PcmAlawDecode(void * pOutBuf, size_t * pOutSamples, const void * pInBuf, const size_t nInSamples);
int PcmMulawDecode(void * pOutBuf, size_t * pOutSamples, const void * pInBuf, const size_t nInSamples);

static const uint8_t m_a2mu[128] =
{
        1,    3,    5,    7,    9,    11,    13,    15,
        16,    17,    18,    19,    20,    21,    22,    23,
        24,    25,    26,    27,    28,    29,    30,    31,
        32,    32,    33,    33,    34,    34,    35,    35,
        36,    37,    38,    39,    40,    41,    42,    43,
        44,    45,    46,    47,    48,    48,    49,    49,
        50,    51,    52,    53,    54,    55,    56,    57,
        58,    59,    60,    61,    62,    63,    64,    64,
        65,    66,    67,    68,    69,    70,    71,    72,
        73,    74,    75,    76,    77,    78,    79,    79,
        80,    81,    82,    83,    84,    85,    86,    87,
        88,    89,    90,    91,    92,    93,    94,    95,
        96,    97,    98,    99,    100,    101,    102,    103,
        104,    105,    106,    107,    108,    109,    110,    111,
        112,    113,    114,    115,    116,    117,    118,    119,
        120,    121,    122,    123,    124,    125,    126,    127
};

static const uint8_t m_mu2a[128] =
{
        1,    1,    2,    2,    3,    3,    4,    4,
        5,    5,    6,    6,    7,    7,    8,    8,
        9,     10,    11,    12,    13,    14,    15,    16,
        17,    18,    19,    20,    21,    22,    23,    24,
        25,    27,    29,    31,    33,    34,    35,    36,
        37,    38,    39,    40,    41,    42,    43,    44,
        46,    48,    49,    50,    51,    52,    53,    54,
        55,    56,    57,    58,    59,    60,    61,    62,
        64,    65,    66,    67,    68,    69,    70,    71,
        72,    73,    74,    75,    76,    77,    78,    79,
        81,    82,    83,    84,    85,    86,    87,    88,
        89,    90,    91,    92,    93,    94,    95,    96,
        97,    98,    99,    100,    101,    102,    103,    104,
        105,    106,    107,    108,    109,    110,    111,    112,
        113,    114,    115,    116,    117,    118,    119,    120,
        121,    122,    123,    124,    125,    126,    127,    128
};

static int16_t ALaw2Linear(uint8_t _nSample)
{
        int nVal;
        int nSeg;

        _nSample ^= 0x55;
        nVal = (_nSample & QUANT_MASK) << 4;
        nSeg = (((unsigned int)(_nSample)) & SEG_MASK) >> SEG_SHIFT;
        switch (nSeg) {
        case 0:
                nVal += 8;
                break;
        case 1:
                nVal += 0x108;
                break;
        default:
                nVal += 0x108;
                nVal <<= nSeg - 1;
        }
        return (int16_t)((_nSample & SIGN_BIT) ? nVal : -nVal);
}

static int16_t MuLaw2Linear(uint8_t _nSample)
{
        int nVal;

        _nSample = ~_nSample;
        nVal = ((_nSample & QUANT_MASK) << 3) + BIAS;
        nVal <<= (((unsigned int)(_nSample)) & SEG_MASK) >> SEG_SHIFT;
        return (int16_t)((_nSample & SIGN_BIT) ? (BIAS - nVal) : (nVal - BIAS));
}

static int Decode(void *_pOutBuf, size_t *_pOutSamples, const void *_pInBuf, const size_t _nInSamples, 
                      int16_t (*_pDecFunc)(uint8_t))
{
        if (_pOutBuf == NULL || _pOutSamples == NULL || _pInBuf == NULL || _nInSamples == 0) {
                return -1;
        }
        int16_t *pDst = (int16_t *)_pOutBuf;
        const uint8_t *pSrc = (uint8_t *)_pInBuf;
        size_t i = 0;
        for (i = 0; i < _nInSamples; i++) {
                *(pDst++) = _pDecFunc(*(pSrc++));
        }
        *_pOutSamples = _nInSamples * 2;
        return -1;
}

int PcmAlawDecode(void * pOutBuf, size_t * pOutSamples, const void * pInBuf, const size_t nInSamples)
{
        return Decode(pOutBuf, pOutSamples, pInBuf, nInSamples, ALaw2Linear);
}

int PcmMulawDecode(void * pOutBuf, size_t * pOutSamples, const void * pInBuf, const size_t nInSamples)
{
        return Decode(pOutBuf, pOutSamples, pInBuf, nInSamples, MuLaw2Linear);
}

